### ISSUE TITLE
Problem drawing text

### ISSUE THREAD

**Issue by labtox** [Microsoft/Win2D#349]
Hello,

I have a problem when drawing text with a custom font that has glyphs in the range 0-31. The font has a 'capital letter A' at index 27 and a 'small letter a' at index 28. The rest of the glyphs are all in the 32+ range. 
When using 'CanvasDrawingSession.DrawText(String, Single, Single, ICanvasBrush, CanvasTextFormat)' to draw text which contains both characters in the sub-32 and the 32+ range, the above mentioned letters 'A' and 'a' do not get drawn, nor is their character width accounted for.

This font is just one example of the many different fonts that are embedded in pdfs, which i need to render. Is there any way I can work around this?

**Comment by clandrew** [comment]
On which font(s) was this happening?

The behavior you're getting is by design. When using DrawText, there is an internal mapping of which characters are considered non-printable. For example, control characters- and, like in your case, code points within the non-printable ASCII range, These are not drawn by default, as it's rare for fonts to have any glyph data for them that an app would want. In most cases, you actually *don't* want control characters to affect spacing at all, much less have any appearance.

Of course, there are exceptions, like when you know you are dealing with a strange font. 

One way to work around this is to use a [custom layout](https://github.com/Microsoft/Win2D-Samples/blob/master/ExampleGallery/Shared/CustomTextLayouts.xaml.cs). A layout checks script analysis for which characters are nonprintable, and skips over them. You could configure your layout to simply not do this, and forcibly draw certain characters (or all characters). Unfortunately, it's a lot of additional code to set that up. 

Another option is to sanitize your source text, changing the non-printable characters to appropriate, equivalent printable characters from that font. If there's no printable character for that font, you can use CanvasTextLayout to apply a different font for certain glyphs.

**Comment by labtox** [comment]
It happens randomly. Embedded PDF fonts are often compressed as much as possible, so this is 'issue' is very common.

I fiugred such a mechanism would be in place.

Is there a certain sample code file which I could use to do so? Or could you give me some directions?
I do have a stripped version of FontForge in place which I could use to shift all glyphs by 32... It would be better though if I could alter the Win2D drawing behaviour.

**Comment by clandrew** [comment]
Just to be totally clear- are the indices you're talking about *Unicode code points*, or *glyph indices in the font*? Could you give an example of the string you pass to DrawText?

It'd be really helpful if you could point us toward PDFs that contain these fonts, for our own test coverage, but that's okay if you don't have them handy.

For custom layout, our [custom layout sample](https://github.com/Microsoft/Win2D-Samples/blob/master/ExampleGallery/Shared/CustomTextLayouts.xaml.cs) describes how to do it.

I don't have a sample handy for sanitizing text, but if you provide more information about your app- preferably, a full repro- I could be able to help.

**Comment by labtox** [comment]
Unfortunately I can't give you our code, but I can provide you a font file with which you can try for yourself.
I use the following code to draw the word 'Agenda' in my COnDraw:

`args.DrawingSession.DrawText(
    "\u001b\u003b\u0032\u004d\u002f\u001c",
    10,
    10,
    Colors.DarkMagenta,
    new CanvasTextFormat {
        FontSize = 30,
        FontFamily = "ms-appx:///Assets/80aa2a8b69a7bcff-OD0754.ttf#OD0754",
        OpticalAlignment = CanvasOpticalAlignment.NoSideBearings,
    } );`

Expected result: 'Agenda'
Actual result: 'gend'
As you can see, both the 'A' and 'a' are sub 32

Here is the font file:
[80aa2a8b69a7bcff-OD0754.zip](https://github.com/Microsoft/Win2D/files/267670/80aa2a8b69a7bcff-OD0754.zip)

And a screenshot of the font file opened in FontForge.
![image](https://cloud.githubusercontent.com/assets/4747477/15315588/d9284e3e-1c18-11e6-8c05-eb87bb7338bc.png)

**Comment by clandrew** [comment]
Thanks for sending this along, I really appreciate it!

I took a look at the font and it appears to be malformed, limiting what I can do with it. It won't open in the built-in Windows Font Viewer, nor some 3rd-party tools I tried, nor is it installable to my system. I'm guessing this is something private and experimental, not a 'real' font that has been shipped somewhere.

I was able to look at some properties of the font using one 3rd-party tool, though, and look at the glyph assignments- there's just a handful of Latin glyphs- and it matches with what FontForge is giving you.

This font's glyph assignments are very much nonconformant, and Win2D is behaving exactly by design here. If I could install this font I suspect we would see the same behavior in other applications as well: latin-script words with dropped-out characters.

Forcing control characters to be printable is something *very* rarely desirable. Even for symbol fonts which use strange glyphs, they will be mapped to visible, non-control-character Unicode ranges.

If this font scenario is a real one, and sanitizing the text is totally impossible, then a custom layout in Win2D is the answer here.
